### PR TITLE
[android][expo-av] Avoid exceptions in AVManager.onHostDestroy()

### PR DESCRIPTION
**Pull request by danmaas**
# Why

In my bare builds that use expo-av for audio, this function throws Java exceptions that cause Google Play's Pre-launch automated tests to report "stability problems" with my app.

# How

- The iteration that calls removeSoundForKey() on all sounds in mSoundMap throws a ConcurrentModificationException, because it tries to remove keys from the map during iteration, which Java does not allow. This patch uses an Iterator to safely clear out the map. (apologies for the copy/paste of the removeSoundForKey() body - that could probably be factored more cleanly).

- unregisterReceiver() seems to be called in cases where the receiver has already been unregistered, or was never registered, causing a Java exception. Added a boolean to track registration status. I am not sure if this is really a fix though. I still need get exceptions from unregisterReceiver(), which I silence here in order to pass Google's pre-launch automated test.

Example stack traces for the two exceptions follow. These came to us via Google Play Store's "Pre-launch report" which is their automated QA process. Note, the exceptions only occurred in about 20% of Google's test runs.

```
10-10 00:59:17.960: E/MonitoringInstr(11279): Exception encountered by: com.badukpop.MainActivity@928dc77. Dumping thread state to outputs and pining for the fjords.
10-10 00:59:17.960: E/MonitoringInstr(11279): java.lang.IllegalArgumentException: Receiver not registered: expo.modules.av.AVManager$1@9def733
10-10 00:59:17.960: E/MonitoringInstr(11279): 	at android.app.LoadedApk.forgetReceiverDispatcher(LoadedApk.java:1095)
10-10 00:59:17.960: E/MonitoringInstr(11279): 	at android.app.ContextImpl.unregisterReceiver(ContextImpl.java:1503)
10-10 00:59:17.960: E/MonitoringInstr(11279): 	at android.content.ContextWrapper.unregisterReceiver(ContextWrapper.java:608)
10-10 00:59:17.960: E/MonitoringInstr(11279): 	at android.content.ContextWrapper.unregisterReceiver(ContextWrapper.java:608)
10-10 00:59:17.960: E/MonitoringInstr(11279): 	at expo.modules.av.AVManager.onHostDestroy(AVManager.java:184)
10-10 00:59:17.960: E/MonitoringInstr(11279): 	at org.unimodules.adapters.react.services.UIManagerModuleWrapper$3.onHostDestroy(UIManagerModuleWrapper.java:170)
10-10 00:59:17.960: E/MonitoringInstr(11279): 	at com.facebook.react.bridge.ReactContext.onHostDestroy(ReactContext.java:237)
10-10 00:59:17.960: E/MonitoringInstr(11279): 	at com.facebook.react.ReactInstanceManager.moveToBeforeCreateLifecycleState(ReactInstanceManager.java:687)
10-10 00:59:17.960: E/MonitoringInstr(11279): 	at com.facebook.react.ReactInstanceManager.onHostDestroy(ReactInstanceManager.java:596)
10-10 00:59:17.960: E/MonitoringInstr(11279): 	at com.facebook.react.ReactInstanceManager.onHostDestroy(ReactInstanceManager.java:610)
10-10 00:59:17.960: E/MonitoringInstr(11279): 	at com.facebook.react.ReactActivityDelegate.onDestroy(ReactActivityDelegate.java:117)
10-10 00:59:17.960: E/MonitoringInstr(11279): 	at com.facebook.react.ReactActivity.onDestroy(ReactActivity.java:70)
10-10 00:59:17.960: E/MonitoringInstr(11279): 	at android.app.Activity.performDestroy(Activity.java:7136)
10-10 00:59:17.960: E/MonitoringInstr(11279): 	at android.app.Instrumentation.callActivityOnDestroy(Instrumentation.java:1158)
10-10 00:59:17.960: E/MonitoringInstr(11279): 	at androidx.test.runner.MonitoringInstrumentation.callActivityOnDestroy(MonitoringInstrumentation.java:177)
10-10 00:59:17.960: E/MonitoringInstr(11279): 	at android.app.ActivityThread.performDestroyActivity(ActivityThread.java:4385)
10-10 00:59:17.960: E/MonitoringInstr(11279): 	at android.app.ActivityThread.handleDestroyActivity(ActivityThread.java:4417)
10-10 00:59:17.960: E/MonitoringInstr(11279): 	at android.app.ActivityThread.-wrap6(ActivityThread.java)
10-10 00:59:17.960: E/MonitoringInstr(11279): 	at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1629)
10-10 00:59:17.960: E/MonitoringInstr(11279): 	at android.os.Handler.dispatchMessage(Handler.java:105)
10-10 00:59:17.960: E/MonitoringInstr(11279): 	at android.os.Looper.loop(Looper.java:156)
10-10 00:59:17.960: E/MonitoringInstr(11279): 	at android.app.Ac
tivityThread.main(ActivityThread.java:6524)
10-10 00:59:17.960: E/MonitoringInstr(11279): 	at java.lang.reflect.Method.invoke(Native Method)
10-10 00:59:17.960: E/MonitoringInstr(11279): 	at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:941)
10-10 00:59:17.960: E/MonitoringInstr(11279): 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:831)
```

```
FATAL EXCEPTION: main
Process: com.coreplane.badukpop.prod, PID: 12110
java.lang.RuntimeException: Unable to destroy activity {com.coreplane.badukpop.prod/com.badukpop.MainActivity}: java.util.ConcurrentModificationException
	at android.app.ActivityThread.performDestroyActivity(ActivityThread.java:4496)
	at android.app.ActivityThread.handleDestroyActivity(ActivityThread.java:4515)
	at android.app.servertransaction.DestroyActivityItem.execute(DestroyActivityItem.java:39)
	at android.app.servertransaction.TransactionExecutor.executeLifecycleState(TransactionExecutor.java:145)
	at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:70)
	at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1816)
	at android.os.Handler.dispatchMessage(Handler.java:106)
	at android.os.Looper.loop(Looper.java:193)
	at android.app.ActivityThread.main(ActivityThread.java:6718)
	at java.lang.reflect.Method.invoke(Native Method)
	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:493)
	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:858)
Caused by: java.util.ConcurrentModificationException
	at java.util.HashMap$HashIterator.nextNode(HashMap.java:1441)
	at java.util.HashMap$KeyIterator.next(HashMap.java:1465)
	at expo.modules.av.AVManager.onHostDestroy(AVManager.java:194)
	at org.unimodules.adapters.react.services.UIManagerModuleWrapper$3.onHostDestroy(UIManagerModuleWrapper.java:170)
	at com.facebook.react.bridge.ReactContext.onHostDestroy(ReactContext.java:237)
	at com.facebook.react.ReactInstanceManager.moveToBeforeCreateLifecycleState(ReactInstanceManager.java:687)
	at com.facebook.react.ReactInstanceManager.onHostDestroy(ReactInstanceManager.java:596)
	at com.facebook.react.ReactInstanceManager.onHostDestroy(ReactInstanceManager.java:610)
	at com.facebook.react.ReactActivityDelegate.onDestroy(ReactActivityDelegate.java:117)
	at com.facebook.react.ReactActivity.onDestroy(ReactActivity.java:70)
	at android.app.Activity.performDestroy(Activity.java:7403)
	at android.app.Instrumentation.callActivityOnDestroy(Instrumentation.java:1306)
	at androidx.test.runner.MonitoringInstrumentation.callActivityOnDestroy(MonitoringInstrumentation.java:177)
	at android.app.ActivityThread.performDestroyActivity(ActivityThread.java:4481)
	... 11 more
```

### CODE REVIEW

File: packages/expo-av/android/src/main/java/expo/modules/av/AVManager.java [hunk 1]
```diff
@@ -181,10 +184,26 @@ public void onHostPause() {
...
   @Override
   public void onHostDestroy() {
-    mContext.unregisterReceiver(mNoisyAudioStreamReceiver);
-    for (final Integer key : mSoundMap.keySet()) {
-      removeSoundForKey(key);
+    if (isRegistered) {
+      try {
+        mContext.unregisterReceiver(mNoisyAudioStreamReceiver);
+        isRegistered = false;
+      } catch (final Exception exception) {
```
Comment (mczernek) [inline packages/expo-av/android/src/main/java/expo/modules/av/AVManager.java]: I don't think we should ignore errors here. Currently, code should not throw an exception. If it does, we want to know that in order ton fix it.

File: packages/expo-av/android/src/main/java/expo/modules/av/AVManager.java [hunk 2]
```diff
@@ -181,10 +184,26 @@ public void onHostPause() {
...
     }
+
+    // removeSoundForKey() on all remaining sounds
+    Iterator<PlayerData> iter = mSoundMap.values().iterator();
+    while(iter.hasNext()) {
+      final PlayerData data = iter.next();
+      iter.remove();
+      if (data != null) {
+        data.release();
+        abandonAudioFocusIfUnused();
```
Comment (mczernek) [inline packages/expo-av/android/src/main/java/expo/modules/av/AVManager.java]: Can't we just call it once, after iteration?

**Review by mczernek — CHANGES_REQUESTED** [review]

**Comment by mczernek** [comment]
Thanks a lot for your work! Please take a look at my comments.

When there are no more suggestions from us, we'll test those changes and merge them if nothing fails.

File: packages/expo-av/android/src/main/java/expo/modules/av/AVManager.java (hunk 1 above)
Comment (danmaas) [reply packages/expo-av/android/src/main/java/expo/modules/av/AVManager.java]: OK, I won't ignore exceptions here.

File: packages/expo-av/android/src/main/java/expo/modules/av/AVManager.java (hunk 2 above)
Comment (danmaas) [reply packages/expo-av/android/src/main/java/expo/modules/av/AVManager.java]: Actually, I see that abandonAudioFocus() is always called at the end of onHostDestroy(), so I think there is no need to call abandonAudioFocusIfUnused() at all. Removed.

File: packages/expo-av/android/src/main/java/expo/modules/av/AVManager.java [hunk 3]
```diff
@@ -89,6 +90,7 @@
   private long mAudioRecorderDurationAlreadyRecorded = 0L;
   private boolean mAudioRecorderIsRecording = false;
   private boolean mAudioRecorderIsPaused = false;
+  private boolean isRegistered = false;
```
Comment (sjchmiela) [inline packages/expo-av/android/src/main/java/expo/modules/av/AVManager.java]: Android Expo codebase tends to follow AOSP code guidelines, so I would be grateful if you would rename this variable to `mIsRegistered`. 🙂

File: packages/expo-av/android/src/main/java/expo/modules/av/AVManager.java [hunk 4]
```diff
@@ -181,10 +184,21 @@ public void onHostPause() {
...
-    for (final Integer key : mSoundMap.keySet()) {
-      removeSoundForKey(key);
+    if (isRegistered) {
+      mContext.unregisterReceiver(mNoisyAudioStreamReceiver);
+      isRegistered = false;
+    }
+
+    // remove all remaining sounds
+    Iterator<PlayerData> iter = mSoundMap.values().iterator();
+    while(iter.hasNext()) {
```
Comment (sjchmiela) [inline packages/expo-av/android/src/main/java/expo/modules/av/AVManager.java]:
```suggestion
    while (iter.hasNext()) {
```

**Review by sjchmiela — COMMENTED** [review]
Thanks from me too! 👏

File: packages/expo-av/android/src/main/java/expo/modules/av/AVManager.java (hunk 3 above)
Comment (danmaas) [reply packages/expo-av/android/src/main/java/expo/modules/av/AVManager.java]: Fixed.

File: packages/expo-av/android/src/main/java/expo/modules/av/AVManager.java (hunk 4 above)
Comment (danmaas) [reply packages/expo-av/android/src/main/java/expo/modules/av/AVManager.java]: Fixed.

**Review by sjchmiela — APPROVED** [review]
👏 👍

**Comment by dorthwein** [comment]
is there a way to use these changes before it gets published?

**Comment by mczernek** [comment]
New SDK will be published soon, but maybe we'll be able to publish `expo-av` before new sdk release. Will let you know if that happens.

**Comment by dorthwein** [comment]
@mczernek i'm presuming `expo-av` gets updated when the new sdk is published as well correct?

Also, thank you for considering publishing the package.

**Comment by mczernek** [comment]
Yes, we usually publish all modules along with SDK.

**Comment by dorthwein** [comment]
@mczernek - sorry to be needy but we've got this bug out in production unfortunately.  While its mitigated, its still firing - is there an ETA on either the expo SDK or a separate publish to npm with this fix?

**Comment by Josh-Burgess** [comment]
> @mczernek - sorry to be needy but we've got this bug out in production unfortunately. While its mitigated, its still firing - is there an ETA on either the expo SDK or a separate publish to npm with this fix?

Probably best to do what we did, which is just to fork the `expo` repo, and publish the `expo-av` portion to a Verdaccio server. Could be waiting a while otherwise, as it's not even in the most recent RC.

**Comment by mczernek** [comment]
Since the SDK36 is taking longer than we expected, I'll publish it before. I will let you know.

**Comment by dorthwein** [comment]
@mczernek THANK YOU!!! @Josh-Burgess - good idea - ideally would prefer not to do that though.  If @mczernek doesn't think it'll be this week for a publish we'll go that route I suppose.  However, I believe the man, the myth, the legend @mczernek will the the epic hero he was born to be and publish early :)

**Comment by dorthwein** [comment]
@mczernek - we've got a solution - release at your leisure - thank you again for all the help and guidance/work.  You as well @Josh-Burgess

**Pull request merged**